### ISSUE TITLE
Problems with function syntax

### ISSUE THREAD

**Issue by ArneBachmann** [evhub/coconut#416]
I just replaced all instances of lambda functions like `() -> ...` by `-> ...` since it seems to be a possible shortcut.
It is compiled, however, to something like `lambda _: ...` which conflicts with many parts of my code, where I use `_` for other things (e.g. the `self` reference or a running variable in comprehensions).

Can you modify the code generation in a way that avoids this problem? Either a more unlikely lambda parameter (`__anonymous__function`) or something automatically generated guaranteed not to conflict?

**Comment by evhub** [comment]
@ArneBachmann This is a feature, not a bug. The intended use case for `-> ...` is to make it exceptionally easy to write simple lambdas. Thus, instead of `x -> x**2`, you can just write `-> _**2`, which is intended to look similar to [Scala's lambda syntax](https://stackoverflow.com/questions/7673545/usage-of-in-scala-lambda-functions). This is documented in the [Lambdas section of the documentation](http://coconut.readthedocs.io/en/master/DOCS.html#lambdas), though looking at it now the documentation of this feature is pretty sparse and should probably be improved.

**Comment by evhub** [comment]
I'm keeping this open with the idea of improving the documentation on this question.

**Comment by ArneBachmann** [comment]
Ah, the confusion comes from the fact that your lambda parameter `_` always has a default value of `None`, which allows to use the lambda with an argument or not.
This way it slipped through my no-args use cases, except for where `_` is used from an outer scope.
Would it be better to enforce using a parameter to avoid such errors?
Coconut -d:
```
lambda _=None: _

>>> a()
a()

>>> a(1)
a(1)

1
x = lambda _: _
x = lambda _: _

>>> x()
x()

Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: <lambda>() missing 1 required positional argument: '_'
>>> x(1)
x(1)

1
```

**Comment by Sitwon** [comment]
No, because you might want a no-argument function for use as a call-back/handler that you can register with an API you don't control.

IMHO, this arises from a misuse of '_'. By convention, you should never assign a value that you care about to '_'. Assigning to '_' signifies that you don't care about the value or that it's unimportant. Hence, you should never use the value stored in '_' as it's expected to be garbage... except in the unique case of these shortcut lambdas.

In Scala's pattern matching, two occurrences of '_' on the LHS need not map to identical values on the RHS (unlike duplicates of any normal identifier, which would cause the match to fail). In which case, you wouldn't have a useful value in '_' at the end of the match. It's not a real/normal identifier.

I would vote for keeping the current behavior and improving the documentation.

**Comment by ArneBachmann** [comment]
But if you don't control the API, how can you be sure that one or zero argument caught by `_` suffices? The API could require two or more arguments... You'd need to know the API anyway, otherwise you could use e.g. a varargs lambda, not a one-param `_`, or am I mistaken?

I can see the point of deliberately choosing this shortcut for the most common use by @evhub, no matter if I like it or not, but I don't get you argument here.

On the point of not using `_` you may very well be in the majority of Python users that agrees to that, but I beg to differ. It love the coding style with `_` instead of `self` or using `_` as an inner variable inside a comprehension, as it makes the code so more short and beautiful, in my opinion.
It's not forbidden by the Python language, but the implementation of default lambda in Coconut makes a deliberate choice to use `_` here, no matter if `_` is already used or not, and that is the mistake, I think.

**Comment by Sitwon** [comment]
So you concede that your use case breaks with the prevalent conventions of the Python community.
And that the feature in Coconut follows the conventions of other functional programming languages (which is Coconut's purpose). 

Yet you're still not satisfied?

It's extremely common to use lambdas for callbacks that take either zero or one argument, so this shortcut works for the two most common use cases of lambdas as it is. For any other use case, Coconut's shortened lambda syntax is already an improvement over vanilla Python.

The only place this is a problem is when you manufacture a situation where your lambda written in the shortcut form is trying to access a variable named `_` in an preceding scope.

```
>>> _ = "Hello"
>>> spam = -> print(_)
>>> spam()
None
```

But once you understand what is happening here (that `-> ...` is a shortcut for `(_=None) -> ...`), you have two obvious and easy solutions available.

```
>>> _ = "hello"
>>> not_ = _  #rename the problematic identifier so it can be seen from the lambda
>>> spam = -> print(not_)
>>> spam()
hello
```
or
```
>>> _ = "hello"
>>> spam = () -> print(_)  # explicitly declare the lambda to take no arguments
>>> spam()
hello
```

I'm curious how you think it would work to give short-cut lambdas a unique identifier name at compile-time without making occurrences of `_` within the body of the lambda ambiguous.

**Comment by ArneBachmann** [comment]
I didn't say any of the Coconut or Python conventions were wrong, and what you lay out in your post is exactly what I described before, is what I do, it's not "manufactured"; I'm also not in immediate need for education on how callbacks work.

The tone of "Yet you're still not satisfied?" sounds pretty hostile to me, as if it was wrong to discuss some language design question and false assumptions on what is the most useful default on a public issue tracker.

I merely said that the Coconut compiler, knowing the closure, could determine if the anonymous optional argument `_` *is* actually used or not, and could make away with it in favor of another variable name not colliding with anything from the outer scope, including a potential `_`.

**Comment by evhub** [comment]
@ArneBachmann @Sitwon I agree that using `_` to store something you actually care about in the outer scope is bad practice, so we shouldn't go out of our way to support that use case. However, I think there's another use case for not compiling to `lambda _=None: ...`, which is enforcing the correct arity. If you write `f = -> 2`, you really don't want to allow `f(1)`, but currently that would work just fine. Thus, I like the suggestion for Coconut to inspect the lambda for use of the implicit `_` variable, and if it is found, compile to `lambda _: ...`, otherwise compile to `lambda: ...`.

**Comment by ArneBachmann** [comment]
That sounds great!
